packages/vite/src/node/plugins/dynamicImportPolyfill.ts

```diff
@@ -16,11 +31,114 @@ export function dynamicImportPolyfillPlugin(config: ResolvedConfig): Plugin {
+          config.logger.warnOnce(
+            `\n'vite/dynamic-import-polyfill' is no longer needed if you target modern browsers`
+          )
+        }
+        if (skip) {
+          return ''
+        }
+        // return a placeholder here and defer the injection to renderChunk
+        // so that we can selectively skip the injection based on output format
+        if (!polyfillString) {
+          polyfillString =
+            `const p = ${polyfill.toString()};` +
+            `${isModernFlag}&&p(${JSON.stringify(resolveModulePath(config))});`
+        }
+        return polyfillString
+      }
+    },
+
+    renderDynamicImport({ format }) {
+      if (skip || format !== 'es') {
+        return null
+      }
+      if (!polyfillString) {
+        throw new Error(
+          `Vite's dynamic import polyfill is enabled but was never imported. This ` +
+            `should only happen when using custom non-html rollup inputs. Make ` +
+            `sure to add \`import "${polyfillId}"\` as the first statement in ` +
+            `your custom entry.`
         )
-        return ''
       }
+      // we do not actually return anything here because rewriting here would
+      // make it impossible to use es-module-lexer on the rendered chunks, which
+      // we need for import graph optimization in ./importAnalysisBuild.
     }
   }
 }
+
+/**
+The following polyfill function is meant to run in the browser and adapted from
+https://github.com/GoogleChromeLabs/dynamic-import-polyfill
+MIT License
+Copyright (c) 2018 uupaa and 2019 Google LLC
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+*/
+
+declare const self: any
+declare const location: any
+declare const document: any
+declare const URL: any
+declare const Blob: any
+
+function polyfill(modulePath = '.', importFunctionName = '__import__') {
+  try {
+    self[importFunctionName] = new Function('u', `return import(u)`)
+  } catch (error) {
+    const baseURL = new URL(modulePath, location)
+    const cleanup = (script: any) => {
```

Comment:
**question:** Is `script` an `HTMLScriptElement`?

Reply:
This was the exact code from the polyfill from google that was removed, so I would prefer to avoid touching it. If we want to improve it, we can do so in another PR.

Reply:
`HTMLScriptElement` would be just only a compile time code addition
Would not touch runtime code :thinking:

Reply:
Let's discuss it in another PR 👍🏼